src/Cardano/Wallet/Primitive.hs

```diff
@@ -283,6 +294,84 @@ restrictedTo (UTxO utxo) outs =
     UTxO $ Map.filter (`Set.member` outs) utxo
 
 
+-- * Slotting
+
+-- | Hard-coded for the time being
+slotsPerEpoch :: Natural
+slotsPerEpoch = 21600
+
+-- | A slot identifier is the combination of an epoch and slot.
+data SlotId = SlotId
+  { epochIndex :: !Word64
+  , slotNumber :: !Word16
+  } deriving stock (Show, Eq, Ord, Generic)
+
+instance NFData SlotId
+
+instance Enum SlotId where
+    toEnum i
+        | i < 0 = error "SlotId.toEnum: bad argument"
+        | otherwise = slotIncr (fromIntegral i) (SlotId 0 0)
+    fromEnum (SlotId e s)
+        | n > fromIntegral (maxBound @Int) =
+            error "SlotId.fromEnum: arithmetic overflow"
+        | otherwise = fromIntegral n
+      where
+        n :: Natural
+        n = fromIntegral e * fromIntegral slotsPerEpoch + fromIntegral s
+
+-- | Add a number of slots to an (Epoch, LocalSlotIndex) pair, where the number
+-- of slots can be greater than one epoch.
+slotIncr :: Natural -> SlotId -> SlotId
+slotIncr n slot = SlotId e s
+  where
+    e = fromIntegral (fromIntegral n' `div` slotsPerEpoch)
+    s = fromIntegral (fromIntegral n' `mod` slotsPerEpoch)
+    n' = n + fromIntegral (fromEnum slot)
+
+-- | @slotDiff a b@ is the number of slots by which @a@ is greater than @b@.
+slotDiff :: SlotId -> SlotId -> Integer
+slotDiff s1 s2 = fromIntegral (fromEnum s1 - fromEnum s2)
+
+-- | Whether the epoch index and slot number are in range.
+isValidSlotId :: SlotId -> Bool
+isValidSlotId (SlotId e s) =
+    e >= 0 && s >= 0 && s < fromIntegral slotsPerEpoch
+
+-- | Calculates which epochs to fetch, given a number of slots, and the start
+-- point. It takes into account the latest block available, and that the most
+-- recent epoch is never available in a pack file.
+epochRange
```

Comment:
This function is not really a general-purpose primitive, so doesn't belong here.
The `tipEpoch - 1` bit is because the current epoch is never packed on cardano-http-bridge.

Reply:
Indeed, I'll move that back onto the HttpBridge :+1: